aten/src/ATen/autocast_mode.cpp

```diff
@@ -108,6 +153,7 @@ enum class CastPolicy : uint8_t {
                      // The wrapper policy is:  append at::kFloat to the args, and redispatch to the
                      // type-aware overload.
   promote, // Run in the widest dtype among several args.
+  user_defined_dtype, // Cast all inputs to user_defined_dtype from input before running the op.
```

Comment:
This feels wrong to me. Suppose that CUDA gets support for bfloat16. It would very reasonably want to unconditionally use `user_defined_dtype` in this situation. So put differently, the cast policy should in fact replace "fp16" with "user_defined_dtype" (but really, I like "lower_precision_fp" or something like that as more descriptive). Then you will want a helper that given a device type gives the user defined dtype *for that device* (and that will unconditionally return Half for CUDA and BFloat16 for CPU.) TBH you don't really need the `thread_local` today, it's easy enough to generalize this code later.

Reply:
`lower_precision_fp` is much more descriptive. Thanks for the suggestions. And I have added the helper to get user defined dtype based on device type input.